scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -164,123 +195,217 @@
-    {:expected-state :expected/completed}))
+    (write-status-to-datomic compute-cluster status)
+    (sandbox/aggregate-exit-code (:exit-code-syncer-state compute-cluster) instance-id 143)
+    {:cook-expected-state :cook-expected-state/completed}))
+
+
+(defn mark-pod-completed-and-kill-pod-in-weird-state
+  "This function is writes a completed state to datomic and also deletes a pod in kubernetes.
+  It is unusual (and unique) because it both modifies kubernetes and modifies datomic. It is intended
+  only to be invoked in pods in state :k8s-actual-state/unknown and handle their recovery."
+  [compute-cluster k8s-actual-state-dict]
+  ; TODO: Should mark mea culpa retry
+  (kill-pod-in-weird-state compute-cluster (pod-has-just-completed compute-cluster k8s-actual-state-dict) k8s-actual-state-dict))
 
 (defn process
-  "Visit this pod-name, processing the new level-state. Returns the new expected state. Returns
+  "Visit this pod-name, processing the new level-state. Returns the new cook expected state. Returns
   empty dictionary to indicate that the result should be deleted. NOTE: Must be invoked with the lock."
-  [{:keys [api-client existing-state-map expected-state-map name] :as kcc} ^String pod-name]
-  (loop [{:keys [expected-state] :as expected-state-dict} (get @expected-state-map pod-name)
-         {:keys [synthesized-state pod] :as existing-state-dict} (get @existing-state-map pod-name)]
+  [{:keys [api-client k8s-actual-state-map cook-expected-state-map name] :as compute-cluster} ^String pod-name]
+  (loop [{:keys [cook-expected-state] :as cook-expected-state-dict} (get @cook-expected-state-map pod-name)
+         {:keys [synthesized-state pod] :as k8s-actual-state-dict} (get @k8s-actual-state-map pod-name)]
     (log/info "In compute cluster" name ", processing pod" pod-name ";"
-              "expected:" (prepare-expected-state-dict-for-logging expected-state-dict) ","
-              "existing:" (prepare-existing-state-dict-for-logging existing-state-dict))
-    ; TODO: We added an :expected/starting state to the machine, to represent when a pod is starting. We map instance.status/unknown to that state
-    ; The todo is to add in cases for [:expected/starting *] for those other states.
+              "cook-expected:" (prepare-cook-expected-state-dict-for-logging cook-expected-state-dict) ","
+              "k8s-actual:" (prepare-k8s-actual-state-dict-for-logging k8s-actual-state-dict))
+    ; We should have the cross product of
+    ;      :cook-expected-state/starting :cook-expected-state/running :cook-expected-state/completed :cook-expected-state/killed :missing
+    ; and
+    ;      :pod/waiting :pod/running :pod/succeeded :pod/failed :pod/unknown :missing
+    ;
+    ; for a total of 30 states.
+    ;
+    ; The only terminal expected states are :cook-expected-state/completed and :missing
+    ; The only terminal pod states are :pod/succeeded :pod/failed and :missing. We also treat :pod/unknown as a terminal state.
```

Comment:
Don't these two sentences contradict each other?

> The only terminal pod states are :pod/succeeded :pod/failed and :missing.

vs.

> We also treat :pod/unknown as a terminal state.

Reply:
No. We don't know if :pod/unknown is a terminal state for kubernetes. We treat it as a terminal state as far as the state machine is concerned, and force a retry at the cook level. The phrasing is very deliberate.

Reply:
Can we add your explanation above? It's really helpful in understanding what the phrasing means.